feat(chatbot-backend): implemented sendMessage mutation that creates …

erik-tsogt opened the pull request:
…a message object in mongodb

Implemented sendMessage mutation that creates a message object. The mutation is jest tested.

DuurenbayarUlz reviewed (commented):
main ee rebase hiigeed, update hiigeerei

DuurenbayarUlz commented on apps/L1US/chatbot/backend/src/schemas/common.schema.ts:
```diff
@@ -9,11 +9,48 @@ export const typeDefs = gql`
+
+  type Query {
+    sampleQuery: String!
+    getMessagesForConversation(conversationID: ID!): [Message]
```
not in the scope of the issue

DuurenbayarUlz commented on apps/L1US/chatbot/backend/src/schemas/common.schema.ts:
```diff
@@ -9,11 +9,48 @@ export const typeDefs = gql`
+  type User {
+    id: ID!
+    userName: String!
+    email: String!
```
not in the scope

DuurenbayarUlz commented on apps/L1US/chatbot/backend/src/resolvers/mutations/send-message-mutation.ts:
```diff
@@ -0,0 +1,20 @@
+    const response = `Echo: ${query}`;
+
+    const newMessage = await MessageModel.create({
+        id: uuidv4(),
```
id's are automatically generated

DuurenbayarUlz commented on apps/L1US/chatbot/backend/src/models/message.model.ts:
```diff
@@ -0,0 +1,19 @@
+
+export interface IMessage extends Document {
+  id: string;
+  chatID: string; 
```
camelCase naming is not consistent

erik-tsogt commented:
@DuurenbayarUlz, I removed the out-of-scope code from the common schema and fixed the redundant ID generation in the Mongoose model. I also updated the mutation and the test.

erik-tsogt commented:
Description
- **Input Validation**: The resolver checks whether the input object is provided. If missing, the mutation throws an error indicating that input is required
- **Input Extraction**: destructures the input obj to get chatId and query fields.
- **Response Generation**: generates a response by prefixing input query with "Echo: ". This currently serves as a placeholder response and will be replaced by the actual logic later on.
- **Database Operation**: Creates a new message document in MongoDB with the provided chatId, query, response, and a timestamp (Now).
- **Return Value**: returns the newly created document, which conforms to the GraphQl Message type (defined in common.schema)

Changes introduced:
- Removed redundant id generation from the message mongoose model.
- Wrote Jest unit test in sendMessage.spec.ts

How to serve, preview and test:
- **To serve**: yarn nx serve chatbot-backend
- **To preview:** yarn nx preview chatbot-backend
- **To test**: yarn test chatbot-backend

erik-tsogt commented:
To manually test:

- Navigate to http://localhost:4200/api/graphql which displays the Apollo Explorer to test mutation
- Select the mutation with its corresponding fields.
- Select the arguments (input:) with their corresponding fields. 
- Add test values and run the mutation.

DuurenbayarUlz commented on apps/L1US/chatbot/backend/src/models/message.model.ts:
```diff
@@ -0,0 +1,17 @@
+const messageSchema = new Schema({
+  chatId: { type: String, required: true },
+  query: { type: String, required: true },
+  response: { type: String, required: true },
```
use { timestamps: true }

DuurenbayarUlz commented on apps/L1US/chatbot/backend/src/resolvers/mutations/send-message-mutation.ts:
```diff
@@ -0,0 +1,18 @@
+        chatId,
+        query,
+        response,
+        timestamp: new Date()
```
when timestamps enabled no need to add timestamp field

DuurenbayarUlz commented on apps/L1US/chatbot/backend/src/schemas/common.schema.ts:
```diff
@@ -8,12 +8,36 @@ export const typeDefs = gql`
+  }
+    
+  type CreateThreadInput {
+    userId: ID!
```
not in the scope

DuurenbayarUlz commented on apps/L1US/chatbot/backend/src/schemas/common.schema.ts:
```diff
@@ -8,12 +8,36 @@ export const typeDefs = gql`
   }
+    
+  type Conversation {
+    userId: ID!
```
not in the scope

DuurenbayarUlz commented on package.json:
```diff
@@ -153,6 +153,7 @@
     "tailwindcss-animate": "1.0.7",
     "toast": "^0.5.4",
     "tslib": "2.3.0",
+    "uuid": "^11.1.0",
```
uninstall the uuid library

DuurenbayarUlz commented on package.json:
```diff
@@ -120,7 +120,7 @@
     "lucide-react": "^0.474.0",
     "moment": "^2.30.1",
-    "mongoose": "8.2.2",
+    "mongoose": "^8.10.1",
```
do we need to update?

DuurenbayarUlz commented on apps/L1US/chatbot/backend/src/resolvers/mutations/send-message-mutation.ts:
```diff
@@ -0,0 +1,17 @@
+import { MutationResolvers } from '../../generated';
+import { MessageModel } from '../../models';
+
+export const sendMessage: MutationResolvers['sendMessage'] = async(_, { input }) => {
```
I updated the TDD in figma, make sure to take a look

DuurenbayarUlz commented on apps/L1US/chatbot/backend/src/resolvers/mutations/send-message-mutation.ts:
```diff
@@ -0,0 +1,17 @@
+    if (!input) {
+        throw new Error('Input is required');
+    }
+    const { chatId, query } = input;
```
since Messages are dependent on conversationID, it would be better to implement createConversation first.

DuurenbayarUlz commented on package.json:
```diff
@@ -153,7 +153,6 @@
     "tailwindcss-animate": "1.0.7",
     "toast": "^0.5.4",
     "tslib": "2.3.0",
-    "validator": "^13.12.0",
```
?
